File: concrete/src/main/java/pcb/uwu/amount/finance/GBP.java
package pcb.uwu.amount.finance;

import org.jetbrains.annotations.NotNull;
import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.unit.finance.CurrencyUnit;

import java.math.BigDecimal;
import java.math.MathContext;

import static pcb.uwu.utils.UnitAmountUtils.dividedByScalar;
import static pcb.uwu.utils.UnitAmountUtils.minusAmount;
import static pcb.uwu.utils.UnitAmountUtils.multipliedByScalar;
import static pcb.uwu.utils.UnitAmountUtils.plusAmount;

public class GBP extends Money {

	public static final CurrencyUnit CURRENCY = CurrencyUnit.of("GBP");

	// region constructor

	public GBP(Number value) {
		super(value, CURRENCY);
	}

	public GBP(String value) {
		super(value, CURRENCY);
	}

	public GBP(BigDecimal value) {
		super(value, CURRENCY);
	}

	public GBP(BigDecimalAmount amount) {
		super(amount, CURRENCY);
	}

	// endregion

	// region implement UnitAmount

	public GBP plus(GBP other) {
		return new GBP(getAmount().plus(other.getAmount().getValue()));
	}

	public GBP minus(GBP other) {
		return new GBP(getAmount().minus(other.getAmount().getValue()));
	}

	@NotNull
	@Override
	public GBP plus(@NotNull UnitAmount<CurrencyUnit> other) {
		return new GBP(plusAmount(this, other));
	}

	@NotNull
	@Override
	public Money minus(@NotNull UnitAmount<CurrencyUnit> other) {
		return new GBP(minusAmount(this, other));
	}

	@Override
	public GBP times(BigDecimal other, MathContext mathContext) {
		return new GBP(multipliedByScalar(this, other, mathContext));
	}

	@Override
	public GBP div(BigDecimal other, MathContext mathContext) {
		return new GBP(dividedByScalar(this, other, mathContext));
	}

	// endregion
}

File: concrete/src/main/java/pcb/uwu/amount/base/currency/GBP.kt
package pcb.uwu.amount.base.currency

import pcb.uwu.amount.base.Currency
import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.base.CurrencyUnit
import pcb.uwu.unit.base.CurrencyUnit.Companion.GBP

class GBP : Currency
{
    @JvmOverloads
    constructor(amount: Number, magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, GBP)

    @JvmOverloads
    constructor(value: String, magnitude: Magnitude = NATURAL)
            : super(value, magnitude, GBP)

    // region UnitAmount

    override fun plus(currency: UnitAmount<CurrencyUnit>) =
        GBP(this.amount + (currency to GBP).amount)

    override fun minus(currency: UnitAmount<CurrencyUnit>) =
        GBP(this.amount - (currency to GBP).amount)

    override fun times(scalar: Number) =
        GBP(this.amount * scalar)

    override fun div(scalar: Number) =
        GBP(this.amount / scalar)

    // endregion
}